[NoahKoh] iP

NoahKoh opened the pull request:
# Luna

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Luna frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/NoahKoh/ip).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :wink:

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

NoahKoh committed:
Level 0

Rename, Greet, Exit

NoahKoh committed:
Level 1

Echo commands

NoahKoh committed:
Level 2

Add, List

NoahKoh committed:
Level 3

Mark as Done/Undone

NoahKoh committed:
Level 4

Tracking ToDos, Events, Deadlines

NoahKoh committed:
TextUiTesting

Automated Text UI Testing

NoahKoh committed:
Level 5

Handle Errors. Deal with errors such as incorrect inputs

NoahKoh committed:
Level 6

Delete. Add support for deleting tasks from the list

NoahKoh committed:
A-Enums

Used Enumerations

NoahKoh committed:
A-Enums

Used Enumerations

NoahKoh committed:
Merge branch 'master' of https://github.com/NoahKoh/ip

NoahKoh committed:
Revert "Merge branch 'master' of https://github.com/NoahKoh/ip"

This reverts commit 91f4bbc51fc62e5b68c059c15a69e1c0f08750f4, reversing
changes made to 77cda80b97d3a3b8380cc040e3e101ea83db24e3.

NoahKoh committed:
Add Level 7

- Save the tasks in the hard disk automatically whenever the task list changes
- Load the data from the hard disk when the chatbot starts up

NoahKoh committed:
Merge branch 'branch-Level-7'

NoahKoh committed:
Add Level 8

- chatbot now understands dates and times

NoahKoh committed:
Merge branch 'branch-Level-8'

NoahKoh committed:
Add A-MoreOOP

- Refactor the code to extract out closely related code as classes

NoahKoh committed:
Merge branch 'A-MoreOOP'

NoahKoh committed:
Add A-Packages

- Organize the classes into one java package

NoahKoh committed:
Merge branch 'branch-A-Packages'

NoahKoh committed:
Merge branch 'add-gradle-support'

NoahKoh committed:
Update mainClass in build.gradle

NoahKoh committed:
Add A-JUnit

- Use JUnit to test public methods of some classes

NoahKoh committed:
Merge branch 'branch-A-JUnit'

NoahKoh committed:
Update build.gradle

NoahKoh committed:
Add A-JavaDoc

- Add header comments to classes/methods

NoahKoh committed:
Add Level 9

- Give users a way to find a task by searching for a keyword in the task description

NoahKoh committed:
Improve coding standards

NoahKoh committed:
Merge branch 'branch-A-JavaDoc'

NoahKoh committed:
Merge branch 'branch-A-CodingStandard'

NoahKoh committed:
Merge branch 'branch-Level-9'

NoahKoh committed:
Add checkstyle

- Ensure code complies with the style rules

NoahKoh committed:
Merge branch 'branch-A-CheckStyle'

NoahKoh committed:
Add Level 10

- Add a GUI to the chatbot. Use the JavaFX technology to implement the GUI.

NoahKoh committed:
Merge branch 'branch-Level-10'

NoahKoh committed:
fix minor checkstyle

jamesberyl commented on src/main/java/luna/Deadline.java:
```diff
@@ -0,0 +1,36 @@
+ */
+public class Deadline extends Task {
+
+    protected LocalDate by;
```
You could consider using a clearer variable name such as dueDate or deadlineDate.

jamesberyl commented on src/main/java/luna/Luna.java:
```diff
@@ -0,0 +1,176 @@
+                    if (inputParts.length < 2) {
+                        return "The description of an event cannot be empty.";
+                        //throw new LunaException("The description of an event cannot be empty.");
+                    }
+                    String[] remainingInput1 = inputParts[1].split("/from", 2);
+                    if (remainingInput1.length < 2) {
+                        return "Event must include a '/from'";
+                        //throw new LunaException("Event must include a '/from'");
+                    }
+                    String description = remainingInput1[0].trim();
+                    String[] remainingInput2 = remainingInput1[1].split("/to", 2);
+                    if (remainingInput2.length < 2) {
+                        return "Event must include a '/to'";
+                        //throw new LunaException("Event must include a '/to'");
+                    }
+                    String from = remainingInput2[0].trim();
+                    String to = remainingInput2[1].trim();
+                    try {
+                        Event task = new Event(description, from, to);
+                        response = taskList.addTask(task);
+                    } catch (DateTimeParseException e) {
+                        return "Invalid date format. Use yyyy-MM-dd.";
+                        //System.err.println("Invalid date format. Use yyyy-MM-dd.");
+                    }
+                } else {
+                    return "Invalid command";
+                    //throw new LunaException("Invalid command");
+                }
+                response += "\n" + "Now you have " + taskList.getTaskList().size() + " tasks in the list.";
+            }
+        } catch (NumberFormatException e) {
+            return "Need a valid Task number";
+            //System.err.println("Need a valid Task number");
+        } catch (IndexOutOfBoundsException e) {
+            return "Task number not in list";
+            //System.err.println("Task number not in list");
+        }
+        storage.save(taskList.getTaskList());
+        return response;
+    }
```
I feel that this getResponse method is a bit too long and complex. You could consider breaking it into smaller, focused helper methods that handle the different commands. 
After that you could use a switch statement to call the respective methods for the respective commands.

jamesberyl commented on src/main/java/luna/Luna.java:
```diff
@@ -0,0 +1,176 @@
+     * @param args Command-line arguments.
+     */
+    public static void main(String[] args) {
+        /*
+        String logo = " ____        _        \n"
+                + "|  _ \\ _   _| | _____ \n"
+                + "| | | | | | | |/ / _ \\\n"
+                + "| |_| | |_| |   <  __/\n"
+                + "|____/ \\__,_|_|\\_\\___|\n";
+        System.out.println("Hello from\n" + logo);
+         */
```
You could consider removing unnecessary code that is not being used. I noticed in a few other places as well.

jamesberyl commented on src/main/java/luna/TaskList.java:
```diff
@@ -0,0 +1,116 @@
+        return "Nice! I've marked this task as done:\n" + task.toString();
+    }
+
+    /**
+     * Marks the task at the specified index as not done and returns a message indicating the task
+     * has been marked as not done.
+     *
+     * @param index The index of the task to mark as not done.
+     * @return A string message indicating the task has been marked as not done.
```
Good consistent use of JavaDocs throughout. Keep it up!

jamesberyl commented on src/main/java/luna/Task.java:
```diff
@@ -0,0 +1,63 @@
+ */
+public class Task implements Serializable {
+    protected String description;
+    protected boolean isDone;
```
Good naming of boolean variable.

jamesberyl reviewed (commented):
Overall, LGTM, just a few nits to fix. I found your code easy to read for most parts. Good use of OOP principles as well. Good job.

brvndonkoh commented on src/main/java/luna/Deadline.java:
Can we use a more descriptive variable name instead of by? It may not be intuitive for those unfamiliar with the project. Perhaps something like deadLine or dueDate?

brvndonkoh commented on src/main/java/luna/Event.java:
Can we use a more descriptive variable name? Perhaps something like startDate and endDate?

brvndonkoh reviewed (commented):
Overall, the usage of JavaDocs is consistent and clear. The code follows proper conventions, and variable naming is generally good, except for the highlighted ones that could be more descriptive. Method names are also well-chosen. Keep up the great work Noah!

NoahKoh committed:
Add Assertions

- Use assert feature to document important assumption that should hold at various points in the code.

NoahKoh committed:
Merge pull request #1 from NoahKoh/branch-A-Assertions

Add Assertions

NoahKoh committed:
Improve Code Quality

The getResponse method is currently handling multiple command cases with a lengthy if-else statement, making it harder to read and maintain. Variable naming conventions and unnecessary comments further reduce readability.

This needs to change to improve the readability, maintainability, and modularity of the code, making it easier to understand and manage.

Move the command handling logic into a new CommandHandler class and replace the if-else statements with a switch case. Improve variable names for clarity and remove unnecessary comments.

This approach encapsulates the command logic within its own class, reducing the complexity of the getResponse method and adhering to coding standards. It also incorporates feedback from the peer review process.

By refactoring in this manner, the codebase becomes more modular, easier to read, and simpler to maintain, thereby improving overall code quality.

NoahKoh committed:
Merge pull request #2 from NoahKoh/branch-A-CodeQuality

Improve Code Quality

NoahKoh committed:
Add Help extension

- Add a help command to display all available command that the user can use

NoahKoh committed:
Merge branch 'branch-C-Help'

NoahKoh committed:
Add UserGuide

NoahKoh committed:
Merge branch 'branch-A-UserGuide'

NoahKoh committed:
Update README.md

NoahKoh committed:
Update README.md for Ui image link
